### PR TITLE
dnsdist: Fix short IOs over TCP

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

`TCPIOHandler::tryRead()` and `TCPIOHandler::tryWrite()` are supposed to update the position and read/write the remaining number of bytes based on the initial total number minus the position, so that the caller can just pass the exact same parameters on short reads/writes until it succeeds or an exception is raised, without having to care about updating the position and number of bytes to read/write.
<strike>Untested yet.</strike>

Supersedes #7971.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/tcpiohandler.hh [hunk 1]
```diff
@@ -208,9 +208,8 @@ public:
       return d_conn->tryRead(buffer, pos, toRead);
     }
```
Comment (omoerbeek) [inline pdns/tcpiohandler.hh]: The cases `toRead == pos` (will cause `::read` to rerturn `0`) and `pos > toRead` (sign wrap) should be handled.

File: pdns/tcpiohandler.hh [hunk 2]
```diff
@@ -243,9 +241,8 @@ public:
       return d_conn->tryWrite(buffer, pos, toWrite);
     }
```
Comment (omoerbeek) [inline pdns/tcpiohandler.hh]: same here

**Comment by rgacogne** [comment]
Added some checks suggested by @omoerbeek and fixed the existing one, which compared the size of the buffer to the requested size, not the requested size minus the existing position.

**Comment by Habbie** [comment]
>  Untested yet.

It fixes the problem in my test environment!

**Comment by rgacogne** [comment]
I forgot to update my comment but I actually tested it in the meantime! I reproduced the issues with short reads and short writes with master, and confirmed that this PR fixed them.
I'll try to write a test for these cases, since it's already the second time I manage to introduce a bug in that exact spot.

**Comment by omoerbeek** [comment]
I approved this, but now it looks to me like not all cases of tryRead were covered.

File: pdns/tcpiohandler.hh [hunk 3]
```diff
@@ -224,28 +223,29 @@ public:
...
+      throw std::out_of_range("Calling tryWrite() with a too small buffer (" + std::to_string(buffer.size()) + ") for a write of " + std::to_string(toWrite - pos) + " bytes starting at " + std::to_string(pos));
+    }
     if (d_conn) {
       return d_conn->tryWrite(buffer, pos, toWrite);
     }
 
-    size_t sent = 0;
     do {
-      ssize_t res = ::write(d_socket, reinterpret_cast<char*>(&buffer.at(pos)), toWrite - sent);
+      ssize_t res = ::write(d_socket, reinterpret_cast<char*>(&buffer.at(pos)), toWrite - pos);
```
Comment (cmouse) [inline pdns/tcpiohandler.hh]: are you accidentically removing constness here?

Comment (cmouse) [reply pdns/tcpiohandler.hh]: also, what if res < 0?

File: pdns/tcpiohandler.hh [hunk 4]
```diff
@@ -192,25 +192,24 @@ public:
...
     }
 
     if (d_conn) {
       return d_conn->tryRead(buffer, pos, toRead);
     }
 
-    size_t got = 0;
     do {
-      ssize_t res = ::read(d_socket, reinterpret_cast<char*>(&buffer.at(pos)), toRead - got);
+      ssize_t res = ::read(d_socket, reinterpret_cast<char*>(&buffer.at(pos)), toRead - pos);
```
Comment (cmouse) [inline pdns/tcpiohandler.hh]: what if res < 0?

File: pdns/tcpiohandler.hh (hunk 3 above)
Comment (omoerbeek) [reply pdns/tcpiohandler.hh]: res < 0 case handled below

Comment (cmouse) [reply pdns/tcpiohandler.hh]: uhm, how?

Comment (omoerbeek) [reply pdns/tcpiohandler.hh]: line 252

Comment (cmouse) [reply pdns/tcpiohandler.hh]: *sigh*, i got fooled by the review view.

**Comment by rgacogne** [comment]
> Seems a few calls to `tryRead` were missed when converting the 3rd arg. There are 4 calls here,
> line 1045 got converted, 1067 looks right, but 936 and 946 need fixing.

You are mistaking `TCPIOHandler::tryRead()` and a static function named `tryRead()`, which exists because we don't speak DoT toward backends atm and so have no need for the filters abstraction. I have a WIP branch unifying those two.

**Comment by omoerbeek** [comment]
> > Seems a few calls to `tryRead` were missed when converting the 3rd arg. There are 4 calls here,
> > line 1045 got converted, 1067 looks right, but 936 and 946 need fixing.
> 
> You are mistaking `TCPIOHandler::tryRead()` and a static function named `tryRead()`, which exists because we don't speak DoT toward backends atm and so have no need for the filters abstraction. I have a WIP branch unifying those two.

Ah, right. Did not spot that and jumped to conlcusions. Having two conventions for the third arg *is* confusing...

**Comment by rgacogne** [comment]
Now with regression tests.

**Pull request merged**